internal/app/repository/repository.go

```diff
@@ -0,0 +1,94 @@
+package repository
+
+import (
+	"errors"
+	"sync"
+)
+
+type Repository interface {
+	SaveURL(u *URL)
+	LoadURL(u *URL) (r *URL, err error)
+	IsUniqueShort(s string) bool
+}
+
+type Store struct {
+	mux *sync.Mutex
+	s   map[string]string
+}
+
+func NewStore() *Store {
+	return &Store{
+		mux: &sync.Mutex{},
+		s:   make(map[string]string),
+	}
+}
+
+type URL struct {
+	FullURL  string
+	ShortURL string
+}
+
+var (
+	ErrURLNotFound = errors.New("url not found")
+)
+
+func newErrURLNotFound() error {
+	return ErrURLNotFound
+}
+
+func (store *Store) SaveURL(u *URL) {
+	store.mux.Lock()
+	defer store.mux.Unlock()
+
+	store.s[u.FullURL] = u.ShortURL
```

Comment:
Лучше сохранять наоборот. 
```suggestion
	store.s[u.ShortURL] = u.FullURL
```
Заданный метод применения - это получить по короткому идентификатору полный url и отправить пользователю редирект. Поиск по ключу однозначно быстрее (точнее не медленнее), чем полный перебор значений.

Reply:
я думал об этом, но выбрал именно такое решение по следующей причине: 
если ключом будет короткий урл, то для того что бы при каждом пост запросе вида ya.ru 
не создавать кучу коротких урлов для одного и того же длинного, на мой взгляд, есть 2 пути - 
или поддерживать 2 мапы - что удвоит хранилище, или валидировать каждый входящий урл при пост запросе. я так и делаю тут.
если я применю предложенные изменения - то в этом месте цикл уйдет, а потом либо появится в момент присвоения короткого адреса, либо пропадет валидация на уникальность полного урла.

Reply:
Можно сделать две взаимно направленных мапы: в одной ключом сделать короткий url, во второй - длинный. Но при этом придется озадачиться синхронным и транзакционно безопасным их изменением. Но в данном случае я не буду настаивать на выборе своего варианта, объяснение принято, если реализация начнет мешать - ты просто сам отрефакторишь код.